walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,194 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  /**
+   * Add new TimeRange to ArrayList by either appending it or combining it with
+   * existing TimeRange(s)
+   */
+  private ArrayList<TimeRange> resolveOverlaps(ArrayList<TimeRange> existingConflicts, 
+                                               TimeRange curEvent) {
+    // Check for overlaps between new event and existing unavailabilities
+    for (int curTime = 0; curTime < existingConflicts.size(); curTime++) {
+      // Since array is sorted chronologically, multiple overlaps will be
+      // sequential
+      if (curEvent.overlaps(existingConflicts.get(curTime))) {
+        int firstOverlap = curTime;
+        while (curTime < existingConflicts.size() &&
+               curEvent.overlaps(existingConflicts.get(curTime))) {
+          curTime++;
+        }
+        curTime--;
+
+        // Insert new combined TimeRange in last overlapped position
+        existingConflicts.set(
+            curTime,
+            combineTimeRanges(curEvent, existingConflicts.get(curTime)));
+        // Delete other overlapped entries
+        existingConflicts.subList(firstOverlap, curTime).clear();
+
+        return existingConflicts;
+      }
+    }
+    // No elements in array or no overlaps
+    existingConflicts.add(curEvent);
+
+    // Sort existing unavailabilities by chronological order of start time
+    Collections.sort(existingConflicts, TimeRange.ORDER_BY_START);
+
+    return existingConflicts;
+  }
+
+  /**
+   * Return a single TimeRange spanning both input TimeRanges
+   */
+  private TimeRange combineTimeRanges(TimeRange t1, TimeRange t2) {
+    int newStart = t1.start() < t2.start() ? t1.start() : t2.start();
+    int newEnd = t1.end() > t2.end() ? t1.end() : t2.end();
+    int newDuration = newEnd - newStart;
+    return TimeRange.fromStartDuration(newStart, newDuration);
+  }
+
+  /**
+   * Return the opposite of input TimeRange ArrayList
+   */
+  private ArrayList<TimeRange> invertTimeRange(ArrayList<TimeRange> t) {
+    if (t.size() < 1) {
+      t.add(TimeRange.WHOLE_DAY);
+      return t;
```

Comment:
This is pretty strange... in this special case, it looks like you're going to modify `t` in-place, but below you return a new time range. I think you should return a new time range in all cases.